### DISCUSSION TITLE
集群情况下，需要每台机器都映射一个端口号出来连接么，

### DISCUSSION THREAD

**Discussion by jhyever** [milvus-io/milvus#38496] [Announcement]
![image](https://github.com/user-attachments/assets/feed9bff-ec5c-45a6-beea-2285c1a146e8)
比如集群机器有三台，A、B、C服务器，集群部署起来running后，A服务器开出来一个映射端口1001，B、C服务器需要映射端口出来么，如果不用映射的话，A服务器如果挂了，怎么连接B、C机器呢，如果开启的话，怎么个机制呢
假如这三台机器A、B、C 用Java代码连接的时候，怎么连接呢，地址端口号怎么配置，如何设置高可用，有具体的案例配置么

**Comment by xiaofan-luan** [comment]
只有proxy需要暴露端口，一般通过loadbalancer绑定proxy的端口来实现

**Reply by jhyever** [reply]
假如这三台机器A、B、C  用Java代码连接的时候，怎么连接呢，地址端口号怎么配置，如何设置高可用，有具体的案例配置么

**Reply by yhmo** [reply]
@jhyever 
参考这个：https://github.com/milvus-io/milvus-sdk-java/issues/1006
假设你有多个milvus server，可以用Java SDK定义一个MilvusMultiServiceClient，MilvusMultiServiceClient继承自MilvusClient，具备大部分功能接口
```
ServerAddress serverAddress = ServerAddress.newBuilder().withHost("address1").withPort(port1).build();
ServerAddress serverSlaveAddress = ServerAddress.newBuilder().withHost("address2").withPort(port2).build();
MultiConnectParam connectParam =  MultiConnectParam.newBuilder().withHosts(Arrays.asList(serverAddress, serverSlaveAddress)).build();
MilvusClient client = new MilvusMultiServiceClient(connectParam);
client.search()
```

MilvusMultiServiceClient内部会定时(间隔10秒)去检查各个server的healthy接口，healthy=ok的则认为是可服务，从可服务的server中选取一个作为主服务。
当你调用写操作接口时(createCollection/dropCollection/insert/upsert/delete等)，MilvusMultiServiceClient会对每一个server调用该接口。
当你调用读操作接口时(search/query/get等)，MilvusMultiServiceClient仅调用主服务的接口。

我的建议是：MilvusMultiServiceClient适用于你有多个服务，并且数据已经ready并且一致，不需要再写，只需查询的场合。如果要对多个server进行写操作，那你就得注意有的server启启停停造成各个server中的数据不一致的问题了。

这里是针对MilvusMultiServiceClient的测试用例，你可以参考里面的写法：
https://github.com/milvus-io/milvus-sdk-java/blob/master/sdk-core/src/test/java/io/milvus/client/MilvusMultiClientDockerTest.java

**Reply by xiaofan-luan** [reply]
> 假如这三台机器A、B、C 用Java代码连接的时候，怎么连接呢，地址端口号怎么配置，如何设置高可用，有具体的案例配置么

通常情况下，你需要使用ALB或者自己部署nginx，代理多个proxy。
绝大多数情况，一个proxy就足够了，除非你的请求很多或者网络流量很多
